enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+related limits) that apply to a single control-plane node cluster with no
+workers also apply globally across all additional added workers. If a large
+amount of workers/pods is desired the user would have to re-install the cluster
+as a regular multi control-plane node cluster.
+
+- Deal with clusters that have less nodes than they had during day-1
+installation. The rationale for this being a non-goal is that it seems that
+removing the "initial" nodes of a cluster causes a lot of issues even in a
+regular, multi-node installation. So dealing with similar issues in single
+control-plane node clusters is also out of scope for this enhancement. For
+example, installing a cluster with 3 control-plane nodes and 3 workers results
+in unschedulable control-plane nodes. Scaling down the worker machine-sets from
+3 to 0 post-installation does not magically make the control-plane nodes
+schedulable, and as a result a lot of infrastructure workloads fail to schedule
+and operators degrade. User intervention is required to fix this.
+
+- Deal with expansion of the single-node control-plane by adding more
+control-plane nodes
+
+- Deal with expansion of clusters that have been installed before the
+implementation of this enhancement (if possible, their expansion may be
+addressed by an upgrade followed with documentation outlining the steps that
+the user has to take to enable expansion, e.g. patch the default
+`IngressController` to have node selector targetting the master pool rather
+than the worker pool).
+
+- Adjust the baremetal platform to support single control-plane node
+installations. The baremetal platform solves the "floating ingress" issue by
+using virtual IP addresses/keepalived.
+
+- Deal with users who want their ingress traffic to not go through the single
+control-plane node.
+
+- Deal with non-cloud, non-"none" platforms such as baremetal, vSphere, etc.
+
+## Proposal
+
+- Create a new `IngressPlacement` API parameter.
+
+- Make sure worker node ignition files are generated even in bootstrap-in-place
+single control-plane node "none"-platform installation.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I want to install a single control-plane
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+Introduce a new field in the Ingress config CR
+(`config.openshift.io/v1/ingresses`) called `IngressPlacement`.
+
+### Implementation Details/Notes/Constraints
+
+This new field will have one of these values - `ControlPlane` or `Workers`. 
+
+In addition, continue to allow the `.spec.replicas` and `.spec.nodePlacement`
+parameters in `operator.openshift.io/v1/ingresscontrollers` CRs to be omitted,
+but change the defaulting behavior for these fields.
```

Comment:
I don't think changing the defaulting behaviour is something that we can do. Typically this would be considered a breaking change as existing consumers may be relying on the existing defaults. Is this considered to be a workload API or a configuration API? If the number of users of the API is sufficiently small then changing the defaults can be possible but it needs to be considered the impact of the change of this default

Edit: Having read further, it's unclear to me what we mean by defaulting, it would be useful to clarify if we are updating the spec of the object during the defaulting or whether this is controller side defaulting only

Reply:
>Is this considered to be a workload API or a configuration API? If the number of users of the API is sufficiently small then changing the defaults can be possible but it needs to be considered the impact of the change of this default

This approach was inspired by @Miciah's  openshift/api#1103 - I guess defaults are an API guarantee and this is truly a breaking change, but would that mean that the linked PR is also a breaking change?

I personally took "optional" omitted parameters in a CR to mean the user saying "I don't have an opinion on this, do whatever makes sense" and if you look at it that way it's not really a breaking change... I'm torn on that.

What if we instead limited the scope of this new API field to affect the defaulting behavior of just the router-default IngressController ? 

>Edit: Having read further, it's unclear to me what we mean by defaulting, it would be useful to clarify if we are updating the spec of the object during the defaulting or whether this is controller side defaulting only

It's definitely a controller thing - the controller creates the initial IngressController if there isn't one already, and its behavior while doing so according to the new API. And it should also initialize those parameters if the installer creates an IngressController with those parameters omitted